app/models/game.rb

```diff
@@ -29,29 +29,210 @@ def populate_board!
+
+  # Return false: no player_piece is 'checking' the opponent king
+    false
+  end
+  
+  def ends_by_checkmate?(color)
+  # Return false immediately unless the game is in check
+    return false unless in_check? color
+  # Return false if opponent can capture the piece(s) causing check
+    return false if opponent_can_capture_piece_causing_check? color
+  # Return false if opponent can block the piece(s) causing check
+    return false if opponent_can_block_piece_causing_check? color
+  # Return false if opponent king can escape check
+    return false if checked_king_can_escape?
+
+    true
+  end
+  
+  def opponent_can_capture_piece_causing_check?(color)
+    @remaining_opponent_pieces.each do |opponent_piece|
+    # Skip this iteration if the opponent_piece is a pawn that couldn't capture the piece causing check
+      if opponent_piece.type.eql? 'Pawn'
+        next if ( opponent_piece.x_coordinate - @x_of_threat ).abs != 1 || ( opponent_piece.y_coordinate - @y_of_threat ).abs != 1
+      end
+      
+    # Check if it would be valid for the opponent_piece to move to the position of the piece causing check
+      if opponent_piece.valid_move? @x_of_threat, @y_of_threat
+      # Check if the opponent_piece wouldn't obstructed while attempting to move to the position of the piece causing check
+        unless opponent_piece.is_obstructed? @x_of_threat, @y_of_threat
+        # If both criteria are met, at least one opponent_piece can capture the piece causing check
+          return true
+        end
+      end
+    end
+
+  # Return false: no opponent_piece can capture the player piece causing check
+    false
+  end
+  
+  def opponent_can_block_piece_causing_check?(color)
+  # Immediately return false if the player piece causing check is a Knight, which can't be blocked
+    return false if @player_piece_causing_check.type.eql? 'Knight'
+  
+  # Remove the opponent king from the remaining_opponent_pieces array, since it can't be involved in blocking
+    @remaining_opponent_pieces.delete_if { |remaining_opponent_piece| remaining_opponent_piece.type.eql? 'King' }
+  
+  # Store the coordinate sets between the player piece causing check and the opponent king
+    in_between_coordinates = set_coordinates_between_piece_causing_check_and_opponent_king
+    
+  # Iterate through coordinate sets between piece causing check, and opponent king
+    in_between_coordinates.each do |x_of_threat, y_of_threat|
+      @remaining_opponent_pieces.each do |opponent_piece|
+      # Check if it would be valid for the opponent_piece move to the specified
+      # coordinate set between the piece causing check and opponent king
+        if opponent_piece.valid_move? x_of_threat, y_of_threat
+        # Check if the opponent_piece wouldn't obstructed while attempting to move
+        # to the specified coordinate set between the piece causing check 
+          unless opponent_piece.is_obstructed? x_of_threat, y_of_threat
```

Comment:
Again, you can save additional space here by using 'next if' instead of 'unless'